driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -228,187 +184,356 @@ public int getMaxWireVersion() {
+            if (skipReleasingServerResourcesOnClose) {
+                serverCursor = null;
             }
-            endOperationInProgress();
 
-            if (result.getResults().isEmpty()) {
-                callback.onResult(null, null);
+            if (getServerCursor() != null) {
+                executeWithConnection(((connection, t) -> {
+                    assertTrue(connection != null || t != null);
+                    if (t != null) {
+                        // guarantee that regardless of exceptions, `serverCursor` is null and client resources are released
+                        serverCursor = null;
+                        releaseClientResources();
+                    } else {
+                        releaseServerAndClientResources(connection);
+                        releaseServerResources(connection);
+                        connection.release();
+                    }
+                }));
             } else {
-                callback.onResult(result.getResults(), null);
+                releaseClientResources();
             }
         }
-    }
 
-    private CommandCursorResult<T> initFromCommandCursorDocument(
-            final ServerAddress serverAddress,
-            final String fieldNameContainingBatch,
-            final BsonDocument commandCursorDocument) {
-        CommandCursorResult<T> cursorResult = new CommandCursorResult<>(serverAddress, fieldNameContainingBatch, commandCursorDocument);
-        cursor.set(cursorResult.getServerCursor());
-        count.addAndGet(cursorResult.getResults().size());
-        postBatchResumeToken = cursorResult.getPostBatchResumeToken();
-        if (LOGGER.isDebugEnabled()) {
-            LOGGER.debug(format("Received batch of %d documents with cursorId %d from server %s", cursorResult.getResults().size(),
-                    cursorResult.getCursorId(), cursorResult.getServerAddress()));
+        void checkCorruptedConnection() {
+            // if `pinnedConnection` is corrupted, then we cannot kill `serverCursor` via such a connection
+            AsyncConnection localPinnedConnection = pinnedConnection;
+            if (localPinnedConnection != null) {
+                skipReleasingServerResourcesOnClose = true;
+            }
         }
-        return cursorResult;
-    }
 
-    private void logQueryResult(final CommandCursorResult<T> result) {
-        LOGGER.debug(format("Received batch of %d documents with cursorId %d from server %s", result.getResults().size(),
-                result.getCursorId(), result.getServerAddress()));
-    }
+        void executeWithConnection(final AsyncCallableWithConnection callable) {
+            assertTrue(state != State.IDLE);
+            if (pinnedConnection != null) {
+                callable.call(assertNotNull(pinnedConnection).retain(), null);
+            } else {
+                assertNotNull(connectionSource).getConnection((connection, t) -> {
+                    if (t instanceof MongoSocketException) {
```

Comment:
This seems in the wrong branch, since `checkCorruptedConnection` is a no-op if `pinnedConnection == null`.

Reply:
Fixed in #401

Reply:
It's not fixed there.  I added another comment in #401 about it though